File: matlab/physiomon_channels.m
%
% physiomon_channels
%
%     contain the channel class. a channel contains informaation abouth the channel but
%     also 2 tempory buffers to push and get data
%    
%         - the device writes data to the channel
%         - the display and storage engine reads data from the channel
%
%     this class makes use of the <cyclicbuffer> class
%
% modifications
%   31-jan-2024   JM    initial version
%   12-feb-2024   JM    now derivated from handle class (pointer logic)

classdef physiomon_channels < handle

  properties (GetAccess = public)
    m_buffers = []
    m_lengthInSeconds = [];
  end

  methods
  
    %% constructor

    function obj = physiomon_channels(nsec)
      
      % constructor of the class, some properties are initialised

      obj.m_lengthInSeconds = nsec;

    end

    %% configure

    function configure(obj,mySettings,myDevice)

      % getting the information that is in instance of the classes <settings> and <device>
      % this method configures the classes buffers and set some properties
      %
      %     syntax : configure(obj,settings,device)
      %
      % with <settings> and instance of the settings class and <device> an instance of the
      % device class

      myChannels = mySettings.getChannels();
      numChan = length(myChannels);
      obj.m_buffers = repmat(struct('sampleRate',[],"display",[],'store',[]),numChan,1);

      % perform for all channels
      
      for iChan = 1:numChan
      
        source = myChannels(iChan).source;
        analogIn = myDevice.getAnalogChannels();

        numAnalog = length(analogIn);
        for iAnalog = 1:numAnalog

          % create the buffers if the source is found 
          
          if ((source+1) == iAnalog)

            sampleRate = analogIn(iAnalog).sampleRate;
            buflen = obj.m_lengthInSeconds * sampleRate;
          
            obj.m_buffers(iChan).sampleRate = sampleRate;
            obj.m_buffers(iChan).display = cyclicbuffer(buflen,'double');
            obj.m_buffers(iChan).store = cyclicbuffer(buflen,'double');

          end

          % Note, other types should be implemented 
        
        end
      end
    end

    %% readDisplay

    function data = readDisplay(obj,channel)
    
      % readDisplay reads the data from the <display> buffer for the appropriate channel.
      % Note, altough this is a read routine, indices in the buffers are updated and so it
      % is necessary to return the instance of the class
      %
      %     syntax : [obj,data] = readDisplay(obj,channel)
      %
      % with <channel> the channel number and <data> a vector with returned data

      data = obj.m_buffers(channel).display.read();
      
    end

    %% readStore
   
    function data = readStore(obj,channel)
    
      % readStore reads the data from the <store> buffer for the appropriate channel.
      % Note, altough this is a read routine, indices in the buffers are updated and so it
      % is necessary to return the instance of the class
      %
      %     syntax : [obj,data] = readStore(obj,channel)
      %
      % with <channel> the channel number and <data> a vector with returned data

      data = obj.m_buffers(channel).store.read();
      
    end

    %% writeData

    function writeData(obj,channel,data)
    
      % the method <writeData> writes <data> to the buffers (display and store) of channel
      % <channel>
      %
      %     syntax : writeData(obj,channel,data)
      %
      % with <channel> the appropriate channel and <data> the data to be written (vector)

      obj.m_buffers(channel).display.write(data);
      obj.m_buffers(channel).store.write(data);

    end

    %% getSampleRate

    function rate = getSampleRate(obj,channel)

      % this method returns the sanplerate for the desired channel
      %
      %     syntax : rate = getSampleRate(obj,channel)
      %
      % with <channel> the desired number and <rate> the returned samplerate
    
      rate = obj.m_buffers(channel).sampleRate;
    end


  end

end

File: python/physiomon_channels.py
#
# physiomon_channels.py
#
#       contains the channels and buffers classes
#
#   modifications
#     18-dec-2023  JM   initial version
#     11-feb-2024  JM   renamed to <physiomon_channels>
#     16-feb-2024  JM   fater implementation of cyclic buffer 

TYPE_ANALOG_IN = 1

class buffer() :

  def __init__(self,length) :
    
    self.m_len = length
    self.m_indxWrite = 0
    self.m_indxRead = 0
    self.m_data = [0] * length

    return

  # read
  #
  #     read the data from the buffer and update the readIndx 
   
  def read(self) :

    data = []
    
    if (self.m_indxWrite > self.m_indxRead) :
      data = self.m_data[self.m_indxRead:self.m_indxWrite]

    if (self.m_indxWrite < self.m_indxRead) :  
      data = self.m_data[self.m_indxRead:] + self.m_data[:self.m_indxWrite]
    
    self.m_indxRead = self.m_indxWrite
  
    #-jm data = []
    #-jm print(" indices old  : %d %d" % (self.m_indxWrite,self.m_indxRead))
    #-jm 
    #-jm while self.m_indxRead != self.m_indxWrite :
    #-jm 
    #-jm  data.append(self.m_data[self.m_indxRead])
    #-jm  
    #-jm  self.m_indxRead = self.m_indxRead + 1
    #-jm  if self.m_indxRead >= self.m_len :
    #-jm    self.m_indxRead = 0
 
    return data
  
  # write
  #
  #     writes the samples in the circular buffer and updates the writeIndex

  def write(self,data) :
   
    ntal = len(data)

    if self.m_indxWrite + ntal <= self.m_len :
      self.m_data[self.m_indxWrite:(self.m_indxWrite + ntal)] = data
      self.m_indxWrite += ntal
    
    else :
      split = self.m_len - self.m_indxWrite
      self.m_data[self.m_indxWrite:] = data[:split]
      self.m_data[:(ntal-split)] = data[split:]
      self.m_indxWrite = ntal - split

    # for dataSample in data :
    #
    #  self.m_data[self.m_indxWrite] = dataSample   
    #  self.m_indxWrite = self.m_indxWrite + 1
    #  if self.m_indxWrite >= self.m_len :
    #    self.m_indxWrite = 0
        
    return
  

class physiomon_channels() :

  # constructor
  #
  #     intialisation of the object, sets the length of the circular buffers

  def __init__(self,nrsec) :
    
    self.m_lengthBuffer = nrsec
    self.m_buffers = []
    return

  # configure
  #
  #     get the settings and the device info to configure the appropriate buffers
    
  def configure(self,settings,device) :

    # for every channel
    
    for i in range(len(settings.m_channels)) :

      tmpBuffer = {"display" : [], "store" : [], "sampleRate" : 0}  
      
      # find the type and perform action depending on the type, currently analog_In is only
      # supported 

      if (settings.m_channels[i]["type"] == TYPE_ANALOG_IN) :

        for analogChan in device.m_analogIn :
          channels = analogChan["channels"]
          try :
            
            indx = channels.index(i)
            tmpBuffer["sampleRate"] = analogChan["sampleRate"]
            tmpBuffer["display"] = buffer(int(self.m_lengthBuffer * tmpBuffer["sampleRate"]))
            tmpBuffer["store"] = buffer(int(self.m_lengthBuffer  * tmpBuffer["sampleRate"]))

            self.m_buffers.append(tmpBuffer.copy())

          except ValueError :
            pass

    return
  
  # readData
  #
  #     reads the data from the internal buffers

  def readData(self,channel) :

    data = self.m_buffers[channel]["display"].read()
    return data
  
  # resetBuffers
  #
  #     resets the buffers

  def resetBuffers(self) :
    return
  
  # writeData
  #
  #     write the data to the internal buffers

  def writeData(self,channel,data) :

    # write to both buffers

    self.m_buffers[channel]["display"].write(data)
    self.m_buffers[channel]["store"].write(data)

    return
  
  # getSampleRate
  #
  #     returns the sampleRate for the desired channel

  def getSampleRate(self,channel) :

    return self.m_buffers[channel]["sampleRate"]
  
  